### PR TITLE
Original paper MNIST implementation

### PR DESCRIPTION
**Pull request by NeelayS**
Not sure what kind of tests to add. If you could add a sample one, I'll add the rest.

### CODE REVIEW

File: KD_Lib/Distilling the Knowledge in a Neural Network/train.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        for (data, label) in train_loader:
+
+            data, label = data.to(device), label.to(device)
+            optimizer.zero_grad()
+
+            soft_label = F.softmax(large_model(data)/temp)
+            out = distil_model(data)
+            soft_out = F.softmax(out/temp)
+
+            loss = (1 - distil_weight) * F.cross_entropy(out, label) + (distil_weight) * loss_fn(soft_label, soft_out)
```
Comment (Het-Shah) [inline KD_Lib/Distilling the Knowledge in a Neural Network/train.py:47]: here in loss_fn for example you can have options like `MSELoss()` and `kl_div()`

**Comment by NeelayS** [comment]
Okay, I'll update it to take the hypeparameters and model sizes as CL arguments.

File: tests/test_KD_Lib.py [hunk 2]
```diff
@@ -14,7 +14,7 @@ def generate_numbers():
     See more at: http://doc.pytest.org/en/latest/fixture.html
     """
 
-    return random.sample(range(100),10)
+    return random.sample(range(100), 10)
```
[the remark this answers is not in the document]
Comment (NeelayS) [reply tests/test_KD_Lib.py:17]: Sorry, but I'm not quite able to figure what tests to write. Should a test for say one of the models check if the dimensions of the output are correct ? 
If you could you add a sample one, that would be great.

[the remark this answers is not in the document]
Comment (Het-Shah) [reply tests/test_KD_Lib.py:17]: hey, no issues. A basic test would be run your code for 1 epoch, like you run your code using main(), so now maybe rename it to let's say `kd_first()` (this is just an example you can give a meaningful name on your own) and just call this function in this `test_KD_Lib.py`. You can locally run tests using `pytest test_KD_Lib.py`. If you still have doubts let me know.

[the remark this answers is not in the document]
Comment (Het-Shah) [reply tests/test_KD_Lib.py:17]: You can see the current test just do the same thing they are calling function and testing them on some input. Right now your `main()` function doesn't have any parameters so you can just call in there.

[the remark this answers is not in the document]
Comment (Het-Shah) [reply tests/test_KD_Lib.py:17]: no need to add asset statements. just call your function.

[the remark this answers is not in the document]
Comment (NeelayS) [reply tests/test_KD_Lib.py:17]: Oh, okay. Was confused about what to assert. Will do it.

[the remark this answers is not in the document]
Comment (Het-Shah) [reply tests/test_KD_Lib.py:17]: Great! Let me know if you still have any doubts.

**Comment by Het-Shah** [comment]
Hey! Any updates?

**Comment by NeelayS** [comment]
Committed some changes. However, I'm a stuck on an import error.

**Comment by Het-Shah** [comment]
Secondly you'll have to add import statements in your `__init__.py` files. You have not written anything in them. Mosty this is the reason they are not getting imported properly. 

For example, in file `KD_Lib/orignal/__init__.py` you should include from 
`KD_Lib.original.mnist import mnist`, and similarly
for the file `KD_Lib/__init__.py` try including
`from KD_Lib.original import mnist`.

**Comment by Het-Shah** [comment]
Try this should help

**Comment by NeelayS** [comment]
Among others now , VSCode shows this as a problem as well - 
`Module 'torch' has no 'flatten' member`
What does this mean ?

**Comment by Het-Shah** [comment]
Don't worry about build errors now. I'll prolly remove build check.

**Comment by Het-Shah** [comment]
> Among others now , VSCode shows this as a problem as well -
Module 'torch' has no 'flatten' member
What does this mean ?

Where are you getting this?

**Comment by NeelayS** [comment]
In the VSCode problems tab. Must be some issue with VSCode.

**Comment by Het-Shah** [comment]
> In the VSCode problems tab. Must be some issue with VSCode.

Otherwise the code is running right?

**Comment by NeelayS** [comment]
Yes

**Comment by Het-Shah** [comment]
Then its fine.

**Comment by Het-Shah** [comment]
There is still one binary file. Can you please remove that.

File: KD_Lib/original/mnist.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+                        help='(Options:SGD,Adam), (Default:SGD)')
+    parser.add_argument('--teacher_size', default=1200, 
+                        help='Default:1200')
+    parser.add_argument('--student_size', default=800, 
+                        help='Default:800')
+    parser.add_argument('--distill_weight', default=0.7, 
+                        help='Default:0.7')
+    parser.add_argument('--temp', default=20, 
+                        help='Default:20')
+    args = parser.parse_args()
```
Comment (khizirsiddiqui) [inline KD_Lib/original/mnist.py:36]: Can you convert these to function arguments? We won't be primarily using these functions from command line.

File: tests/test_KD_Lib.py [hunk 4]
```diff
@@ -5,30 +5,35 @@
...
-def test_sum_numbers(generate_numbers):
-    """Sample test function for sum_numbers, using pytest fixture."""
+#     our_result = KD_Lib.sum_numbers(generate_numbers)
+#     assert our_result == sum(generate_numbers)
 
-    our_result = KD_Lib.sum_numbers(generate_numbers)
-    assert our_result == sum(generate_numbers)
 
+# def test_max_number(generate_numbers):
+#     """Sample test function for max_number, using pytest fixture."""
```
Comment (khizirsiddiqui) [inline tests/test_KD_Lib.py:30]: You can remove these statements, instead of commenting.

File: KD_Lib/original/train.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+            epoch_loss += loss
+
+        loss_arr.append(epoch_loss)
+        print(f'Epoch {e+1} loss = {epoch_loss}')
+
+    plt.plot(loss_arr)
+    plt.show()
+
+
+def train_student(large_model, distil_model, train_loader, optimizer, loss_fn, epochs=10, temp=20, distil_weight=0.7):
```
Comment (khizirsiddiqui) [inline KD_Lib/original/train.py:29]: line length more than 80 chars, please make the args multiline, styling requirements

File: KD_Lib/original/train.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    large_model.eval()
+    distil_model.train()
+    loss_arr = []
+
+    for e in range(epochs):
+        epoch_loss = 0
+
+        for (data, label) in train_loader:
+
+            data, label = data.to(device), label.to(device)
```
Comment (khizirsiddiqui) [inline KD_Lib/original/train.py:40]: same as before

File: KD_Lib/original/train.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+            optimizer.zero_grad()
+
+            soft_label = F.softmax(large_model(data)/temp)
+            out = distil_model(data)
+            soft_out = F.softmax(out/temp)
+
+            loss = (1 - distil_weight) * F.cross_entropy(out, label) + (distil_weight) * loss_fn(soft_label, soft_out)
```
Comment (khizirsiddiqui) [inline KD_Lib/original/train.py:47]: same as before

**Comment by Het-Shah** [comment]
Also pull the master branch first then make these changes.

**Comment by Het-Shah** [comment]
Remove matplotlib, it isn't a library we have kept in requirements.txt. I don't think we need to use matplotlib.

File: KD_Lib/original/train.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+    model.train()
+    loss_arr = []
+
+    print('Training teacher - \n')
+
+    for e in range(epochs):
+        epoch_loss = 0
+        for (data, label) in train_loader:
+            data, label = data.to(device), label.to(device)
```
Comment (Het-Shah) [inline KD_Lib/original/train.py:15]: Also device isn't defined anywhere. Please make these changes. The build should pass now.

File: KD_Lib/original/evaluate.py [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+
+def eval(model, data_loader):
+
+    model.eval()
+    total = len(data_loader.dataset)
+    correct = 0
+
+    with torch.no_grad():
+        for data, target in data_loader:
+            data, target = data.to(device), target.to(device)
```
Comment (Het-Shah) [inline KD_Lib/original/evaluate.py:12]: here as well.

**Comment by Het-Shah** [comment]
Great merging now.

**Pull request merged**